File: copy_string.py
# Function to copy string (Iterative and Recursive)
import string
from common import tramp, generate_string_repetition


def copy_string(s: str) -> str:
    copy = ''
    for c in s:
        copy += c

    return copy


def copy_string_rec(s: str) -> str:
    def helper(idx: int, acc='') -> str:
        if idx >= len(s):
            return acc
        curr = s[idx]
        return helper(idx+1, acc+curr)
    return helper(0)


def copy_string_rec_gen(s: str) -> str:
    def helper(idx: int, acc='') -> str:
        if idx >= len(s):
            yield acc
        curr = s[idx]
        yield helper(idx+1, acc+curr)
    yield helper(0)


x = 'mystring'
y = copy_string(x)
z = copy_string_rec(x)
print(y == z)

alphabet = string.ascii_lowercase
repeated_alphabet = generate_string_repetition(alphabet)
# print(copy_string_rec(repeated_alphabet))
print(tramp(copy_string_rec_gen, repeated_alphabet))

File: javascript/src/copy_string.ts
// Function to copy string (Iterative and Recursive)
import { Trampoline, tramp, alphabetString } from './global';

function copyStringIter(s: string): string {
    let newString = '';
    for (const c of s) {
        newString += c;
    }
    return newString;
}

function copyStringRec(s: string): string {
    const c = s[0];
    if (c == undefined) {
        return '';
    }
    return c + copyStringRec(s.slice(1));
}

function copyStringTailRec(s: string): string {
    function helper(idx: number, acc = ''): string {
        const c = s[idx];
        if (c == undefined) {
            return acc;
        }
        return helper(idx + 1, acc + c);
    }
    return helper(0);
}

function copyStringThunk(s: string): Trampoline<string> {
    function helper(idx: number, acc = ''): Trampoline<string> {
        const c = s[idx];
        if (c == undefined) {
            return acc;
        }
        return () => helper(idx + 1, acc + c);
    }
    return helper(0);
}

const repeatedAlphabet = alphabetString.repeat(10_000);

console.log(copyStringIter(alphabetString) === copyStringRec(alphabetString)
    && copyStringRec(alphabetString) === copyStringTailRec(alphabetString));

console.log(tramp(copyStringThunk, repeatedAlphabet));